tutorials/crunch-aggregate-sessions.md

```diff
@@ -0,0 +1,296 @@
+
+### AggregateEvents.java
+
+```Java
+package org.kitesdk.examples.tutorials.crunch;
+
+import java.io.Serializable;
+import java.net.URI;
+import java.util.Calendar;
+import java.util.Iterator;
+import java.util.TimeZone;
+import org.apache.crunch.DoFn;
+import org.apache.crunch.Emitter;
+import org.apache.crunch.MapFn;
+import org.apache.crunch.PCollection;
+import org.apache.crunch.Pair;
+import org.apache.crunch.Target;
+import org.apache.crunch.types.avro.Avros;
+import org.apache.crunch.util.CrunchTool;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.util.ToolRunner;
+import org.kitesdk.data.Dataset;
+import org.kitesdk.data.Datasets;
+```
+Kite's CrunchDatasets class methods handle the underlying code required to create source and target datasets for use with Crunch.
+
+```Java
+import org.kitesdk.data.crunch.CrunchDatasets;
+import org.kitesdk.data.event.StandardEvent;
+import org.kitesdk.data.spi.filesystem.FileSystemDatasets;
+import org.kitesdk.examples.demo.event.Session;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class AggregateEvents extends CrunchTool implements Serializable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AggregateEvents.class);
+
+  @Override
+  public int run(String[] args) throws Exception {
+
+    // Turn debug on while in development.
+    getPipeline().enableDebug();
+    getPipeline().getConfiguration().set("crunch.log.job.progress", "true");
+
+```
+
+Step 1. Load the source dataset. In this example, use the `events` dataset from Hive.
+
+```Java
+    Dataset<StandardEvent> eventsDataset = Datasets.load(
+        "dataset:hive:events", StandardEvent.class);
+```
+
+If the dataset is empty, stop.
+
+```Java
+    if (eventsDataset.isEmpty()) {
+      LOG.info("No records to process.");
+      return 0;
+    }
+```
+
+Still going? Good. Create a _parallel collection_ using the `events` dataset as the source. A parallel collection (`PCollection`) is a set of distributed objects that can be processed in parallel.
+
+```Java
+    PCollection<StandardEvent> events = read(
+        CrunchDatasets.asSource(eventsDataset));
+```
+
+Step 2. Process the data.
+a. Create a session key by combining the user ID and session ID.
+b. Group together all events with the same session key.
+c. For each group, create a Session record as an Avro Specific object. 
+
+```Java
+    PCollection<Session> sessions = events
+        .by(new GetSessionKey(), Avros.strings())
+        .groupByKey()
+        .parallelDo(new MakeSession(), Avros.specifics(Session.class));
+
+```
+
+Step 3. Append the derived sessions to the target dataset, defined using the Kite `CrunchDatasets.asTarget` method.
+
+```Java
+    getPipeline().write(sessions,
+        CrunchDatasets.asTarget("dataset:hive:sessions"),
+        Target.WriteMode.APPEND);
+    return run().succeeded() ? 0 : 1;
+  }
+```
+
+The `GetSessionKey` class combines the session ID with the user ID to create a unique session key.
```

Comment:
I think this should note that this is the implementation of the step used above.